Source/plrmsg.cpp

```diff
@@ -110,26 +122,55 @@ void InitPlrMsg()
 void DrawPlrMsg(const Surface &out)
 {
 	int x = 10;
-	int y = 58;
+	int y = PANEL_TOP - 32;
 	int width = gnScreenWidth - 20;
+	int vislines;
+	int chatlines;
 	_plrmsg *pMsg;
+	int lineHeight = 15;
 
-	if (chrflag || QuestLogIsOpen) {
+	if (!talkflag && (chrflag || QuestLogIsOpen)) {
 		x += GetLeftPanel().position.x + GetLeftPanel().size.width;
 		width -= GetLeftPanel().size.width;
 	}
-	if (invflag || sbookflag)
+	if (!talkflag && (invflag || sbookflag))
 		width -= gnScreenWidth - GetRightPanel().position.x;
 
 	if (width < 300)
 		return;
 
+	if (talkflag) {
+		vislines = PMSG_COUNT;
+	} else {
+		vislines = 3;
+	}
+
 	pMsg = plr_msgs;
-	for (int i = 0; i < PMSG_COUNT; i++) {
-		if (pMsg->str[0] != '\0')
-			PrintChatMessage(out, x, y, width, pMsg->str, TextColorFromPlayerId[pMsg->player]);
-		pMsg++;
-		y += 35;
+
+	std::string text = WordWrapString(pMsg->str, width);
+	if (!text.empty()) {
+		chatlines = GetChatLines(width, text);
+	}
+
+	for (int i = 0; i < vislines; i++) {
+		y -= lineHeight * chatlines;
+
+		if (pMsg->str[0] != '\0') {

```

Comment:
Since you removed `ClearPlrMsg` this no longer has the intended meaning of reaching the end of messages.

Reply:
This loop is intended to show all 8 messages on screen when the chat box is open, and to display message slots 0-2 until they reach 10000 ticks when the chat box is closed

Reply:
The problem is that the string is never cleared so you already know if it's empty or not based on what the current counter is at.

Reply:
line 158 is to prevent the game from drawing the transparent box where no lines exist

Reply:
... yes, but you could have used your counter instead.